src/dbnode/storage/series/buffer.go

```diff
@@ -268,65 +385,106 @@ func bucketTick(now time.Time, b *dbBuffer, idx int, start time.Time) int {
 
-	if b.buckets[idx].needsDrain(now, start) {
-		// Rotate the buffer to a block, merging if required
-		result, err := b.buckets[idx].discardMerged()
-		if err != nil {
+	if bucket.needsDrain(now, start) {
+		mergedOutOfOrderBlocks += b.bucketDrain(now, id, start)
+	}
+
+	if bucket.needsReset(now, start) {
+		bucket.resetTo(start, true)
+	}
+
+	if bucket.isStale(now) {
+		b.removeBucket(id.key)
+	}
+
+	return mergedOutOfOrderBlocks
+}
+
+func (b *dbBuffer) bucketDrain(now time.Time, id bucketID, start time.Time) int {
+	mergedOutOfOrderBlocks := 0
+
+	bucket := b.bucketAtIdx(id)
+	// Rotate the buffer to a block, merging if required
+	result, err := bucket.discardMerged()
+	if err != nil {
+		log := b.opts.InstrumentOptions().Logger()
+		log.Errorf("buffer merge encode error: %v", err)
+	} else {
+		if result.merges > 0 {
+			mergedOutOfOrderBlocks++
+		}
+		if !(result.block.Len() > 0) {
 			log := b.opts.InstrumentOptions().Logger()
-			log.Errorf("buffer merge encode error: %v", err)
+			log.Errorf("buffer drain tried to drain empty stream for bucket: %v",
+				start.String())
 		} else {
-			if result.merges > 0 {
-				mergedOutOfOrderBlocks++
+			// If this block was read mark it as such
+			if lastRead := bucket.lastRead(); !lastRead.IsZero() {
+				result.block.SetLastReadTime(lastRead)
 			}
-			if !(result.block.Len() > 0) {
-				log := b.opts.InstrumentOptions().Logger()
-				log.Errorf("buffer drain tried to drain empty stream for bucket: %v",
-					start.String())
-			} else {
-				// If this block was read mark it as such
-				if lastRead := b.buckets[idx].lastRead(); !lastRead.IsZero() {
-					result.block.SetLastReadTime(lastRead)
-				}
+			if b.drainFn != nil {
 				b.drainFn(result.block)
 			}
 		}
-
-		b.buckets[idx].drained = true
 	}
 
-	if b.buckets[idx].needsReset(start) {
-		// Reset bucket
-		b.buckets[idx].resetTo(start)
+	bucket.drained = true
+	bucket.setLastWrite(now)
+	bucket.resetNumWrites()
+	if bucket.isStale(now) {
+		b.removeBucket(id.key)
 	}
 
 	return mergedOutOfOrderBlocks
 }
 
+func (b *dbBuffer) makeBucketRealTime(fromKey xtime.UnixNano, toIdx int) {
+	bucket, ok := b.bucketsNotRealTime[fromKey]
+	if !ok {
+		return
+	}
+
+	bucket.isRealTime = true
+	b.bucketsRealTime[toIdx] = *bucket
+	b.removeBucket(fromKey)
+}
+
 func (b *dbBuffer) Bootstrap(bl block.DatabaseBlock) error {
 	blockStart := bl.StartTime()
 	bootstrapped := false
-	for i := range b.buckets {
-		if b.buckets[i].start.Equal(blockStart) {
-			if b.buckets[i].drained {
+	for i := range b.bucketsRealTime {
+		if b.bucketsRealTime[i].start.Equal(blockStart) {
+			if b.bucketsRealTime[i].drained {
 				return fmt.Errorf(
 					"block at %s cannot be bootstrapped by buffer because its already drained",
 					blockStart.String(),
 				)
 			}
-			b.buckets[i].bootstrap(bl)
+			b.bucketsRealTime[i].bootstrap(bl)
 			bootstrapped = true
 			break
 		}
 	}
+	if bucket, ok := b.bucketsNotRealTime[xtime.ToUnixNano(blockStart)]; !bootstrapped && ok {
```

Comment:
You're gonna do the map lookup here even if `bootstrapped = true` you can probably refactor this to prevent that

Reply:
Actually this whole code path is a little unclear to me. When would we be bootstrapping non-realtime blocks? I think this portion will need some careful thought with how we plan to handle bootstrapping and non-realtime data...

Reply:
Probably worth a discussion in person